challenges/2021-11-16-data-grouping/solutions/javascript/justinmemphis/justinmemphis.js

```diff
@@ -0,0 +1,156 @@
+6. The results of the live coding demo will be added to the GitHub repository as a mob solution. Please feel free to tackle this challenge in your own way/language and submit your solution to the repo using [this solutions guide in our wiki](https://github.com/codeconnector/CodingDojo/wiki#solutions).
+
+*/
+
+function mean(input) {
+	var sum = 0;
+	for (var i = 0; i < input.length; i++) {
+		sum += +input[i][1];
+	};
+	var result = sum / input.length;
+	return +result.toFixed(2);
+};
+
+function median(input) {
+	var numbers = [];
+	for (var i = 0; i < input.length; i++) {
+		numbers.push(+input[i][1]);
+	};
+
+	const sorted = Array.from(numbers).sort((a,b) => a - b);
+	const middle = Math.floor(sorted.length / 2);
+
+	if (sorted.length % 2 === 0) {
+		return (sorted[middle -1] + sorted[middle]) / 2;
+	}
+
+	return sorted[middle];
+};
+
+function higherThan25(input) {
+	let count = 0;
+	for (var i = 0; i < input.length; i++) {
+		if (+input[i][1] > 25) {
+			count++;
+		};
+	};
+	return count;
+};
+
+function twentyFiveToTwentyNine(input) {
+	let count = 0;
+	for (var i = 0; i < input.length; i++) {
+		if (+input[i][1] > 25 && +input[i][1] < 30) {
+			count++;
+		};
+	};
+	return count;
+};
+
+function sortData(inputData, schoolYear, schoolYear2 = 'blah') {
+	let result = inputData.filter(element => element[0] == schoolYear || element[0] == schoolYear2);
+	return result;
+};
+
+function groupData() {
+	// type your solution here
+	fetch('http://justincarter.tech/HS_ACT_Data.txt')
+		.then(response => response.text())
+		.then((responseData) => {
+			const data = responseData.split('\r\n').map((line) => line.split('\t'));
+			//console.log(data);
+
+			let freshmen = sortData(data, 'Freshman');
+			let sophomores = sortData(data, 'Sophomore');
+			let juniors = sortData(data, 'Junior');
+			let seniors = sortData(data, 'Senior');
+			/*
+			console.log('Freshmen: ',freshmen.length,', ',
+				'Sophomores: ',sophomores.length,', ',
+				'Juniors: ',juniors.length,', ',
+				'Seniors: ',seniors.length,', ',
+				'Total: ',data.length);
+			*/
+
+			let sophomoresAndJuniors = sortData(data, 'Sophomore', 'Junior');
+			//console.log('Sophomores and Juniors: ',sophomoresAndJuniors.length);
+
+			console.log(`
+HighSchool Year         Mean        	      Median
+Freshman                ${mean(freshmen)}			${median(freshmen)}
+Sophomores              ${mean(sophomores)}			${median(sophomores)}
+Juniors  		${mean(juniors)}			${median(juniors)}
+Seniors  		${mean(seniors)}			${median(seniors)}
+			`);
+
+			console.log('return the number of scores greater than 25: ',higherThan25(data));
+			console.log('return the number of sophomore scores greater than 25 and less than 30: ',twentyFiveToTwentyNine(sophomores));
+			console.log('return the mean of only the sophomore and junior scores: ',mean(sophomoresAndJuniors));
+		});
+};
+
+groupData();
+
+// module.exports = { groupData };
```

Comment:
You _will_ need this for testing, though. Here are the tests I wrote to confirm that your code correctly calculates all the summaries (which it definitely does, kudos!)

```javascript
const { groupData } = require("./justinmemphis");

test("Freshman Mean and Median", () => {
  const result = groupData();
  expect(result["Freshman Mean"]).toEqual(24.4);
  expect(result["Freshman Median"]).toEqual(23);
});

test("Sophomore Mean and Median", () => {
  const result = groupData();
  expect(result["Sophomore Mean"]).toEqual(25.0);
  expect(result["Sophomore Median"]).toEqual(24);
});

test("Junior Mean and Median", () => {
  const result = groupData();
  expect(result["Junior Mean"]).toEqual(24.6);
  expect(result["Junior Median"]).toEqual(24);
});

test("Senior Mean and Median", () => {
  const result = groupData();
  expect(result["Senior Mean"]).toEqual(25.2);
  expect(result["Senior Median"]).toEqual(25);
});

test("Scores Over 25", () => {
  const result = groupData();
  expect(result["Scores Over 25"]).toEqual(730);
});

test("Sophomore Scores Between 25 and 30", () => {
  const result = groupData();
  expect(result["Sophomore Scores Between 25 and 30"]).toEqual(84);
});

test("Sophomore and Junior Mean", () => {
  const result = groupData();
  expect(result["Sophomore and Junior Mean"]).toEqual(24.7);
});
```

Reply:
Thank you Eric!  I added this and the tests are passing for me.
See commit below.